### ISSUE TITLE
Add support for returning a mapped tensor for multi-valued text matching rank features

### ISSUE THREAD

**Issue by jobergum** [vespa-engine/vespa#29549]
Given 

```
schema doc {

  document doc {
     field chunks type array<string> {}
  }
  field embedding type tensor<float>(p{}, x[384]) {}
```

You can do things like

```
rank-profile example {
  inputs {
      query(q) tensor<float>(x[384])
  }
  function all_sims() {
    expression: cosine_similarity(query(q), attribute(embedding), x)
  }
  first-phase {
      expression: reduce(all_sims, max, p)
  }
}
```
But, the current text-matching features that work with multi-valued fields aggregate element-scores into a single score. 

### Suggestion
We should expose multi-valued rank features that produce a score per element and return a mapped tensor with the dimension `dim` 

For example, in our case where we p:

```
cosine_similarity(query(q), attribute(embedding), x) * elementTextSimilarity(chunks, p) 
```

```
cosine_similarity(query(q), attribute(embedding), x) * bm25(chunks, p) 
```

**Comment by iach1234** [comment]
happy user here, i followed an archived slack message asking, about "per element BM25 scores for an array<string> datatype".

per element BM25 scores would be wonderful.

it might be a fairly common use case in which a long text documents (e.g. reports, papers, court judgments) are stored as multivector. there often tends to be one or two highly relevant chunks in the document, with other chunks being somewhat irrelevant. Per element BM25 be a very indicative signal.

**Comment by dadib** [comment]
I second this would be a very useful feature for longer documents. We would like to use something like this for narrowing a document hit to its best matching parts for dynamic snippets and for context selection for LLM answer generation.

**Comment by bratseth** [comment]
A more general solution could be to be able to apply a lambda function to each array element and get the result as a tensor.

**Comment by jobergum** [comment]
> For us this is a really important feature from a result visualization perspective.

Yes, I think it's important too and an area of improvement, which is why we have this feature issue :) 

> Without this feature we would have to adopt a work-around for the client service, which is essentially to iterate the >paragraphs and find the one with the highest number of bolding tags (eg <hi>), 

Bolding tags would be an imperfect proxy for the score,  but in essence yes. You can also with regular keyword matching use [matched-elements-only](https://docs.vespa.ai/en/reference/schema-reference.html#matched-elements-only) to eliminate those elements that doesn't match any of the query terms searching the field.

**Comment by Alexander-Mark** [comment]
We've also recently run into this issue: https://docs.vespa.ai/en/performance/sizing-search.html#scaling-retrieval-size

Our docs occasionally have a very large number of elements in an `array<string>` field, which significantly increases the latency when performing a ColBERT max sim reranking operation.

If we could return values for individual elements, we could better control exposure in subsequent reranking phases eg: apply second phase max sim computation on only the top N first phase elements (by bm25 etc)

**Comment by NikeNano** [comment]
This would be great!

**Comment by geirst** [comment]
Elementwise bm25 rank feature for multi-value index fields is available in Vespa 8.478.26:

- Reference doc: https://docs.vespa.ai/en/reference/rank-features.html#rank-scores
- System test for example usage: https://github.com/vespa-engine/system-test/blob/master/tests/search/bm25_feature/test.sd
- Implementation: https://github.com/vespa-engine/vespa/pull/33287
